File: crates/rosette-wasm/src/viewport.rs
//! Viewport state for camera control.
//!
//! Matches rosette-web's coordinate system:
//! - zoom: pixels per world unit (higher = more zoomed in)
//! - offset: screen position of world origin in pixels
//! - screenPos = worldPos * zoom + offset

/// Viewport state for camera control.
///
/// Uses rosette-web's offset-based coordinate system.
#[derive(Debug, Clone, Copy)]
pub struct Viewport {
    /// Screen position of world origin in pixels.
    pub offset_x: f64,
    pub offset_y: f64,

    /// Zoom level (pixels per world unit).
    /// Higher values = more zoomed in.
    pub zoom: f64,

    /// Canvas size in pixels.
    pub width: u32,
    pub height: u32,

    /// Theme: true = dark, false = light.
    pub dark_theme: bool,

    /// Device pixel ratio for HiDPI/retina display support.
    pub dpr: f32,
}

/// Default zoom level: 2^-6 = 0.015625 (matches rosette-web)
const DEFAULT_ZOOM: f64 = 0.015625; // 2^-6

/// Minimum zoom level to prevent grid rendering issues at extreme zoom-out.
const MIN_ZOOM: f64 = 1e-18;

impl Default for Viewport {
    fn default() -> Self {
        Self {
            offset_x: 400.0, // Will be updated when canvas size is known
            offset_y: 300.0,
            zoom: DEFAULT_ZOOM,
            width: 800,
            height: 600,
            dark_theme: true,
            dpr: 1.0,
        }
    }
}

impl Viewport {
    /// Create a new viewport with default settings.
    pub fn new() -> Self {
        Self::default()
    }

    /// Set the offset (screen position of world origin).
    pub fn set_offset(&mut self, x: f64, y: f64) {
        self.offset_x = x;
        self.offset_y = y;
    }

    /// Set the zoom level (pixels per world unit).
    pub fn set_zoom(&mut self, zoom: f64) {
        const MAX_ZOOM: f64 = 3.0;
        self.zoom = zoom.clamp(MIN_ZOOM, MAX_ZOOM);
    }

    /// Set the canvas size and optionally recenter.
    pub fn set_size(&mut self, width: u32, height: u32) {
        self.width = width;
        self.height = height;
    }

    /// Set the theme.
    pub fn set_dark_theme(&mut self, dark: bool) {
        self.dark_theme = dark;
    }

    /// Set the device pixel ratio for HiDPI/retina display support.
    pub fn set_dpr(&mut self, dpr: f32) {
        self.dpr = dpr.max(1.0);
    }

    /// Convert world coordinates to screen coordinates.
    /// screenPos = worldPos * zoom + offset
    pub fn world_to_screen(&self, world_x: f64, world_y: f64) -> (f32, f32) {
        let screen_x = world_x * self.zoom + self.offset_x;
        let screen_y = world_y * self.zoom + self.offset_y;
        (screen_x as f32, screen_y as f32)
    }

    /// Convert screen coordinates to world coordinates.
    /// worldPos = (screenPos - offset) / zoom
    pub fn screen_to_world(&self, screen_x: f64, screen_y: f64) -> (f64, f64) {
        let world_x = (screen_x - self.offset_x) / self.zoom;
        let world_y = (screen_y - self.offset_y) / self.zoom;
        (world_x, world_y)
    }

    /// Get the visible bounds in world coordinates.
    /// Matches rosette-web's getVisibleBounds function.
    pub fn visible_bounds(&self) -> (f64, f64, f64, f64) {
        // startX: -offset.x / zoom
        // endX: (width - offset.x) / zoom
        let start_x = -self.offset_x / self.zoom;
        let end_x = (self.width as f64 - self.offset_x) / self.zoom;
        let start_y = -self.offset_y / self.zoom;
        let end_y = (self.height as f64 - self.offset_y) / self.zoom;
        (start_x, start_y, end_x, end_y)
    }
}

/// GPU-compatible viewport uniform data.
///
/// This struct is designed to be uploaded directly to a GPU uniform buffer.
#[repr(C)]
#[derive(Debug, Clone, Copy, bytemuck::Pod, bytemuck::Zeroable)]
pub struct ViewportUniform {
    /// Screen position of world origin in pixels.
    pub offset: [f32; 2],
    /// Zoom level (pixels per world unit).
    pub zoom: f32,
    /// Theme: 1.0 = dark, 0.0 = light.
    pub theme: f32,
    /// Canvas size in pixels.
    pub size: [f32; 2],
    /// Device pixel ratio for HiDPI/retina display support.
    pub dpr: f32,
    /// Padding for 16-byte alignment.
    pub _padding: f32,
    /// Cell origin in world coordinates, used for crosshair positioning.
    pub crosshair_origin: [f32; 2],
    /// Padding for 16-byte alignment (total struct = 48 bytes).
    pub _padding2: [f32; 2],
}

impl ViewportUniform {
    /// Create from a viewport with the given crosshair origin in world coordinates.
    pub fn from_viewport(vp: &Viewport, crosshair_origin: [f32; 2]) -> Self {
        Self {
            offset: [vp.offset_x as f32, vp.offset_y as f32],
            zoom: vp.zoom as f32,
            theme: if vp.dark_theme { 1.0 } else { 0.0 },
            size: [vp.width as f32, vp.height as f32],
            dpr: vp.dpr,
            _padding: 0.0,
            crosshair_origin,
            _padding2: [0.0, 0.0],
        }
    }
}

impl From<&Viewport> for ViewportUniform {
    fn from(vp: &Viewport) -> Self {
        Self::from_viewport(vp, [0.0, 0.0])
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_world_to_screen_origin() {
        let vp = Viewport {
            offset_x: 400.0,
            offset_y: 300.0,
            zoom: 1.0,
            width: 800,
            height: 600,
            dark_theme: true,
            dpr: 1.0,
        };

        // World origin (0,0) should map to offset position
        let (sx, sy) = vp.world_to_screen(0.0, 0.0);
        assert_eq!(sx, 400.0);
        assert_eq!(sy, 300.0);
    }

    #[test]
    fn test_screen_to_world_roundtrip() {
        let vp = Viewport {
            offset_x: 400.0,
            offset_y: 300.0,
            zoom: 0.5,
            width: 800,
            height: 600,
            dark_theme: true,
            dpr: 1.0,
        };

        let (wx, wy) = vp.screen_to_world(100.0, 200.0);
        let (sx, sy) = vp.world_to_screen(wx, wy);

        assert!((sx - 100.0).abs() < 0.001);
        assert!((sy - 200.0).abs() < 0.001);
    }
}

File: app/src/stores/viewport.ts
import { create } from "zustand";
import { ZOOM_PADDING_PERCENT } from "@/lib/constants";

/**
 * Viewport state for camera control.
 *
 * Matches rosette-web's coordinate system:
 * - zoom: pixels per world unit (higher = more zoomed in)
 * - offset: screen position of world origin in pixels
 * - screenPos = worldPos * zoom + offset
 */
/** Bounds in world coordinates. */
interface WorldBounds {
  minX: number;
  minY: number;
  maxX: number;
  maxY: number;
}

interface ViewportState {
  /** Zoom level (pixels per world unit). Higher = more zoomed in. */
  zoom: number;
  /** Screen position of world origin in pixels. */
  offset: { x: number; y: number };
  /** Whether the viewport has been initialized. */
  initialized: boolean;

  /** Set zoom level. */
  setZoom: (zoom: number) => void;
  /** Zoom by a factor around a screen point. */
  zoomAt: (factor: number, screenX: number, screenY: number) => void;
  /** Pan by a delta in screen pixels. */
  pan: (dx: number, dy: number) => void;
  /** Set offset directly. */
  setOffset: (x: number, y: number) => void;
  /** Reset to default view (origin at screen center). */
  reset: (canvasWidth: number, canvasHeight: number) => void;
  /** Initialize offset to center origin on screen (only runs once). */
  initOffset: (canvasWidth: number, canvasHeight: number) => void;
  /** Zoom to fit the given world bounds with padding.
   * Optional screenCenter overrides where the bounds are centered on screen
   * (defaults to canvas center). Used to account for floating panels. */
  zoomToBounds: (
    bounds: WorldBounds,
    canvasWidth: number,
    canvasHeight: number,
    screenCenter?: { x: number; y: number },
  ) => void;
  /** Zoom to fit all objects, or reset to default if no bounds provided. */
  zoomToFit: (
    bounds: WorldBounds | null,
    canvasWidth: number,
    canvasHeight: number,
    screenCenter?: { x: number; y: number },
  ) => void;
  /** Zoom to fit selected objects (no-op if no bounds provided). */
  zoomToSelected: (
    bounds: WorldBounds | null,
    canvasWidth: number,
    canvasHeight: number,
    screenCenter?: { x: number; y: number },
  ) => void;
  /** Center view on bounds without changing zoom level. */
  centerOnBounds: (
    bounds: WorldBounds,
    canvasWidth: number,
    canvasHeight: number,
    screenCenter?: { x: number; y: number },
  ) => void;
}

/**
 * Grid size: pixels per grid point at zoom=1.
 * Each grid point represents 1 nanometer.
 * Matches rosette-web for consistent scaling.
 */
const GRID_SIZE = 50;

/** Default zoom level. Matches rosette-web: 2^-6 = 0.015625 */
const DEFAULT_ZOOM = Math.pow(2, -6);

/** Zoom bounds. MIN_ZOOM prevents grid rendering issues at extreme zoom-out levels. */
const MIN_ZOOM = 1e-18;
const MAX_ZOOM = 3;

export const useViewportStore = create<ViewportState>((set, get) => ({
  zoom: DEFAULT_ZOOM,
  offset: { x: 0, y: 0 }, // Will be initialized when canvas mounts
  initialized: false,

  setZoom: (zoom) =>
    set({
      zoom: Math.max(MIN_ZOOM, Math.min(MAX_ZOOM, zoom)),
    }),

  zoomAt: (factor, screenX, screenY) => {
    const state = get();
    const newZoom = Math.max(MIN_ZOOM, Math.min(MAX_ZOOM, state.zoom * factor));

    // Calculate world position under cursor: worldPos = (screenPos - offset) / zoom
    const worldX = (screenX - state.offset.x) / state.zoom;
    const worldY = (screenY - state.offset.y) / state.zoom;

    // Calculate new offset to keep world position under cursor
    // screenPos = worldPos * newZoom + newOffset
    // newOffset = screenPos - worldPos * newZoom
    const newOffsetX = screenX - worldX * newZoom;
    const newOffsetY = screenY - worldY * newZoom;

    set({
      zoom: newZoom,
      offset: { x: newOffsetX, y: newOffsetY },
    });
  },

  pan: (dx, dy) =>
    set((state) => ({
      offset: {
        x: state.offset.x + dx,
        y: state.offset.y + dy,
      },
    })),

  setOffset: (x, y) => set({ offset: { x, y } }),

  reset: (canvasWidth, canvasHeight) =>
    set({
      zoom: DEFAULT_ZOOM,
      offset: { x: canvasWidth / 2, y: canvasHeight / 2 },
      initialized: true,
    }),

  initOffset: (canvasWidth, canvasHeight) => {
    const state = get();
    // Only initialize once - use explicit flag instead of checking offset values
    // (offset {0,0} is a valid panned position)
    if (!state.initialized) {
      set({
        offset: { x: canvasWidth / 2, y: canvasHeight / 2 },
        initialized: true,
      });
    }
  },

  zoomToBounds: (bounds, canvasWidth, canvasHeight, screenCenter?) => {
    // Calculate bounds dimensions
    const baseWidth = Math.abs(bounds.maxX - bounds.minX);
    const baseHeight = Math.abs(bounds.maxY - bounds.minY);

    // Use minimum bounds size for points/lines to ensure they're visible
    // 1000 world units = 1 micron, provides reasonable context around the object
    const MIN_BOUNDS_SIZE = 1000;
    const effectiveWidth = Math.max(baseWidth, MIN_BOUNDS_SIZE);
    const effectiveHeight = Math.max(baseHeight, MIN_BOUNDS_SIZE);

    const paddingX = effectiveWidth * ZOOM_PADDING_PERCENT;
    const paddingY = effectiveHeight * ZOOM_PADDING_PERCENT;
    const width = effectiveWidth + paddingX * 2;
    const height = effectiveHeight + paddingY * 2;

    // Center of bounds in world coordinates
    const centerX = (bounds.minX + bounds.maxX) / 2;
    const centerY = (bounds.minY + bounds.maxY) / 2;

    // Calculate zoom to fit bounds in canvas (clamped to valid range)
    const zoomX = canvasWidth / width;
    const zoomY = canvasHeight / height;
    const newZoom = Math.max(MIN_ZOOM, Math.min(zoomX, zoomY, MAX_ZOOM));

    // Calculate offset to center bounds on screen
    // screenPos = worldPos * zoom + offset
    // offset = screenCenter - worldCenter * zoom
    const cx = screenCenter ? screenCenter.x : canvasWidth / 2;
    const cy = screenCenter ? screenCenter.y : canvasHeight / 2;
    const newOffset = {
      x: cx - centerX * newZoom,
      y: cy - centerY * newZoom,
    };

    set({
      zoom: newZoom,
      offset: newOffset,
    });
  },

  zoomToFit: (bounds, canvasWidth, canvasHeight, screenCenter?) => {
    if (bounds) {
      get().zoomToBounds(bounds, canvasWidth, canvasHeight, screenCenter);
    } else {
      // No objects - reset to default view, centering on visible area
      const cx = screenCenter ? screenCenter.x : canvasWidth / 2;
      const cy = screenCenter ? screenCenter.y : canvasHeight / 2;
      set({
        zoom: DEFAULT_ZOOM,
        offset: { x: cx, y: cy },
        initialized: true,
      });
    }
  },

  zoomToSelected: (bounds, canvasWidth, canvasHeight, screenCenter?) => {
    // Only zoom if we have bounds (i.e., there's a selection)
    if (bounds) {
      get().zoomToBounds(bounds, canvasWidth, canvasHeight, screenCenter);
    }
    // No-op if no selection
  },

  centerOnBounds: (bounds, canvasWidth, canvasHeight, screenCenter?) => {
    const state = get();

    // Center of bounds in world coordinates
    const centerX = (bounds.minX + bounds.maxX) / 2;
    const centerY = (bounds.minY + bounds.maxY) / 2;

    // Calculate offset to center bounds on screen (keep current zoom)
    // screenPos = worldPos * zoom + offset
    // offset = screenCenter - worldCenter * zoom
    const cx = screenCenter ? screenCenter.x : canvasWidth / 2;
    const cy = screenCenter ? screenCenter.y : canvasHeight / 2;
    const newOffset = {
      x: cx - centerX * state.zoom,
      y: cy - centerY * state.zoom,
    };

    set({ offset: newOffset });
  },
}));

// Export constants and types for use elsewhere
export { GRID_SIZE, DEFAULT_ZOOM, MIN_ZOOM, MAX_ZOOM };
export type { WorldBounds };

/**
 * When running inside an iframe (embed mode), broadcast viewport changes to
 * the parent frame so the host page can render a synced background dot grid.
 */
if (typeof window !== "undefined" && window.parent !== window) {
  const postViewport = (state: ViewportState) => {
    window.parent.postMessage(
      {
        type: "rosette-viewport",
        zoom: state.zoom,
        offsetX: state.offset.x,
        offsetY: state.offset.y,
      },
      "*",
    );
  };

  // Broadcast on every state change
  useViewportStore.subscribe(postViewport);

  // Also broadcast the initial state so the host page can render immediately
  postViewport(useViewportStore.getState());
}
